### PR TITLE
Add response objects

### PR DESCRIPTION
**Pull request by be9**
The basic idea is to replace `Hash`-based response objects with real objects. This allows to write `response.transactions[0].date` instead of `response['transactions'][0]['date']`. The latter is: a) non-Rubyish, b) too weak and error-prone.

The implementation is based on [`Hashie::Dash`](https://github.com/intridea/hashie#dash), which is a hash with a fixed set of keys. It doesn't introduce any backward incompatibility, since all response objects are derived from `Hash`. The old variant, `response['transactions'][0]['date']`, will still work, but you won't be able to call `response['trasnactions'][0]['dte']`, as the keys are validated.

Apart from adding response objects:

* Added dwolla and apex processor endpoints.
* The code was heavily refactored (introducing `BaseProduct` with some metaprogramming removed much boilerplate).
* Rubocop added as a linter.

### CODE REVIEW

**Comment by hanchang** [comment]
Very much looking forward to seeing this merged into the gem! Having to use hashes is a big pain, thanks for all the hard work!

**Comment by javierjulio** [comment]
@be9 @michaelckelly it would be very helpful to have this. Any chance this will still make its way into the gem? Would you consider a simpler implementation that just generates objects from the hashes instead as that might require less changes?

**Comment by be9** [comment]
@javierjulio sorry for the long wait and thanks for the ping. We expect to finally fix and merge it next week.

**Comment by javierjulio** [comment]
@be9 awesome thanks! You've done great work here. 🙏🏻 Let me know if I can help to get it through. Thanks again! ❤️

File: lib/plaid/products/item.rb [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
     #
-    # Returns a parsed JSON with the new access_token and request id
+    # Returns an InvalidateResponse object with the new access_token and
+    # request id.
     def invalidate(access_token)
-      payload = { access_token: access_token }
-      @client.post_with_auth('item/access_token/invalidate', payload)
+      post_with_auth 'item/access_token/invalidate',
+                     InvalidateResponse,
+                     access_token: access_token
```
Comment (jdudek) [inline lib/plaid/products/item.rb]: Nit: would be nice to be consistent with parens around method calls (we do add parens elsewhere)

File: lib/plaid/products/income.rb [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
     #
-    # Does a POST /income/get call which returns income info for an access_tokeni's item
+    # Does a POST /income/get call which returns income info for an
+    # access_token's item.
     #
-    # access_token - access_token who's item to fetch income for
+    # access_token - access_token whose item to fetch income for.
     #
-    # Returns a parsed JSON of income information
+    # Returns the IdentityResponse object with Income info.
```
Comment (jdudek) [inline lib/plaid/products/income.rb]: `s/IdentityResponse/IncomeResponse/`

File: lib/plaid/products/item.rb (hunk 1 above)
Comment (be9) [reply lib/plaid/products/item.rb]: I went ahead and removed all the parens surrounding post_with_auth :)

File: lib/plaid/products/income.rb (hunk 2 above)
Comment (be9) [reply lib/plaid/products/income.rb]: :zap:

**Comment by be9** [comment]
@jdudek :zap:, also added Dwolla and Apex.

**Pull request merged**